### ISSUE TITLE
[Refunds] After refunding, the app always shows "via manual refund" until the Order is refreshed

### ISSUE THREAD

**Issue by joshheald** [woocommerce/woocommerce-ios#6704]
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

**To Reproduce**
1. Go to an order, paid via a payment gateway which supports automatic refunds (e.g. WCPay)
2. Tap `Issue Refund`, and refund the payment to the original payment method
3. On the order details, note that the subtitle incorrectly says "via manual refund"
4. Pull to refresh the order
5. Note that the subtitle now correctly shows the refund payment method

**Screenshots**

https://user-images.githubusercontent.com/2472348/165113846-6423fa8b-dbc4-41d7-b4ab-95e32852846d.mp4

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by iamgabrielma** [comment]
I took an initial look at this one, and so far the problem seems to be when the check for `hasRefundGateway` in [OrderPaymentDetailsViewModel](https://github.com/woocommerce/woocommerce-ios/blob/b84e421e871501d4486ef6ebce8dec8b4f9d1192/WooCommerce/Classes/ViewModels/Order%20Details/OrderPaymentDetailsViewModel.swift#L130) happens, versus when we actually update the UI.

This check sets the Order refund type to "manual refund" if no automated payment gateway is set in the Order object, we might be triggering it too early as doesn't really matter which payment gateway we use, it always defaults to "manual payment" on the first instance.

```        
let refundType = hasRefundGateway ? order.paymentMethodTitle : NSLocalizedString(
            "manual refund",
            comment: "A manual refund is one where the store owner has given the purchaser alternative funds" +
                " (cash, check, ACH) instead of using the payment gateway to create a refund " +
                "(credit card or debit card was refunded)"
        )
```
For a test Order, when we trigger the refund from the app, `WooCommerce.OrderDetailsViewController.viewModel.refunds` immediately returns `isAutomated = false`, however, after refreshing the View, using `UIRefreshControl`, or if we navigate to a different View, when we go back to the Order Details, the `isAutomated` parameter returns `true` instead (which is the expected value in this case, as payment gateways like Stripe or WCPay allow for automated refunds).

My initial guess is that we have to update or retrieve the Order details within a completion handler in the refund process, so we're sure that all data that will appear in the View is accurate. For example, as @joshheald pointed out, we can use something similar to when we [update the Order asynchronously when collecting a payment](https://github.com/woocommerce/woocommerce-ios/blob/f08f59f7bf0893d8f3989381090a04e0a2d46df6/WooCommerce/Classes/ViewRelated/Orders/Simple%20Payments/Method/SimplePaymentsMethodsViewModel.swift#L231).

**Comment by iamgabrielma** [comment]
After digging a bit more, I confirmed that we're setting the refund type before the refund flow is completed, specifically, the Refund object shows the following when we're calling for the `refundType` initially:

```
Yosemite.Refund	
isAutomated	Bool?	nil	none
createAutomated	Bool?	nil	none
```

Later on within the flow, we synchronize the Order object properly within `OrderDetailsViewController` each time we call `viewWillAppear()` via [syncEverything()](https://github.com/woocommerce/woocommerce-ios/blob/02c0822186500e5abe4b29a588c469c4be4deb8c/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderDetailsViewController.swift#L83), however, as the refund flow appears via a Modal screen, `viewWillAppear()` is not called again until we actually reload the View, which causes that we can't see the refund type just after the refund flow is completed, but this appears only when we reload the View.

The tl;dr would be: We're looking into refreshing the ViewController when the modal screen for the refund flow is completed.

**Comment by joshheald** [comment]
Nice investigation @iamgabrielma 

FYI `isAutomated` is named differently when it comes over the network:

[Here's where we decode it into the Refund model](https://github.com/woocommerce/woocommerce-ios/blob/23d213c373fa2c45f8255e04d3b6270981fc3353/Networking/Networking/Model/Refund/Refund.swift#L74).

[You can see that in the keys, that uses `refunded_payment`](https://github.com/woocommerce/woocommerce-ios/blob/23d213c373fa2c45f8255e04d3b6270981fc3353/Networking/Networking/Model/Refund/Refund.swift#L123) as its property to decode from the JSON.

It's worth getting Proxyman or Charles set up if you don't already and we can look at what's coming over the network and when, which is always useful with this sort of investigation 😊... that way you can take a look at what's happening when we create the refund and when we get it back on a refresh. We'll look in to this together later when we pair, I'm just mentioning it in case you want to get a head start.

**Comment by iamgabrielma** [comment]
Thanks! The session definitely helped a lot to understand the problem better, I'll write down some takes but please feel free to correct me if I got something wrong, there may be a few gaps:

> FYI isAutomated is named differently when it comes over the network:
> [Here's where we decode it into the Refund model](https://github.com/woocommerce/woocommerce-ios/blob/23d213c373fa2c45f8255e04d3b6270981fc3353/Networking/Networking/Model/Refund/Refund.swift#L74).

Right! As you pointed out this is only the case name in the Enum, the value we're actually looking for comes with a different name over the API: `refunded_payment`.  [API reference](https://woocommerce.github.io/woocommerce-rest-api-docs/#order-refund-properties) 

When we [initialize the Refund object](https://github.com/woocommerce/woocommerce-ios/blob/23d213c373fa2c45f8255e04d3b6270981fc3353/Networking/Networking/Model/Refund/Refund.swift#L123), one of the keys we're attempting to decode is `refunded_payment` into `isAutomated`, or default to false if this property is not present in the API request:

`let isAutomated = try container.decodeIfPresent(Bool.self, forKey: .automatedRefund) ?? false`

The problem is that in our tests, the property `refunded_payment` was not present in the POST request, setting this to false on initialization, however, this can be found in the GET request. Examples:

POST endpoint `/wc/v3/orders/$ORDER_ID/refunds&_method=post`

<details>
{
  "data": {
    "id": 6680,
    "date_created": "2022-06-08T09:05:22",
    "date_created_gmt": "2022-06-08T09:05:22",
    "amount": "0.83",
    "reason": ":)",
    "refunded_by": 1,
    "meta_data": [
      {
        "id": 3959,
        "key": "_wcpay_refund_id",
        "value": "re_3L8KoO2Ev1jcPAUs1jPqhXeL"
      }
    ],
    "line_items": [
      {
        "id": 25,
        "name": "TEST Product!",
        "product_id": 6649,
        "variation_id": 0,
        "quantity": -1,
        "tax_class": "",
        "subtotal": "-0.83",
        "subtotal_tax": "0.00",
        "total": "-0.83",
        "total_tax": "0.00",
        "taxes": [],
        "meta_data": [
          {
            "id": 238,
            "key": "_refunded_item_id",
            "value": "24",
            "display_key": "_refunded_item_id",
            "display_value": "24"
          }
        ],
        "sku": "",
        "price": 0.826446,
        "parent_name": null
      }
    ],
    "shipping_lines": [],
    "tax_lines": [],
    "fee_lines": [],
    "_links": {
      "self": [
        {
          "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679/refunds/6680"
        }
      ],
      "collection": [
        {
          "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679/refunds"
        }
      ],
      "up": [
        {
          "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679"
        }
      ]
    }
  }
}

</details>

GET endpoint: `/wc/v3/orders/$ORDER_ID/refunds&_method=get`

<details>
{
  "data": [
    {
      "id": 6680,
      "date_created": "2022-06-08T09:05:22",
      "date_created_gmt": "2022-06-08T09:05:22",
      "amount": "0.83",
      "reason": ":)",
      "refunded_by": 1,
      "refunded_payment": true,
      "meta_data": [
        {
          "id": 3959,
          "key": "_wcpay_refund_id",
          "value": "re_3L8KoO2Ev1jcPAUs1jPqhXeL"
        }
      ],
      "line_items": [
        {
          "id": 25,
          "name": "TEST Product!",
          "product_id": 6649,
          "variation_id": 0,
          "quantity": -1,
          "tax_class": "",
          "subtotal": "-0.83",
          "subtotal_tax": "0.00",
          "total": "-0.83",
          "total_tax": "0.00",
          "taxes": [],
          "meta_data": [
            {
              "id": 238,
              "key": "_refunded_item_id",
              "value": "24",
              "display_key": "_refunded_item_id",
              "display_value": "24"
            }
          ],
          "sku": "",
          "price": 0.826446,
          "parent_name": null
        }
      ],
      "shipping_lines": [],
      "tax_lines": [],
      "fee_lines": [],
      "_links": {
        "self": [
          {
            "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679/refunds/6680"
          }
        ],
        "collection": [
          {
            "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679/refunds"
          }
        ],
        "up": [
          {
            "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc/v3/orders/6679"
          }
        ]
      }
    }
  ]
}
</details>


**Suggestions and next steps:**

- This is tangential, but we can remove ` ?? false` from [here](https://github.com/woocommerce/woocommerce-ios/blob/23d213c373fa2c45f8255e04d3b6270981fc3353/Networking/Networking/Model/Refund/Refund.swift#L74). `isAutomated` is declared already as an Optional so we don't need to default to `false` if not found.

- Is it expected that the API POST request does not contain `refunded_payment` at this point? This could be a bug or a feature, so we would want either a temporary or a permanent workaround. For the moment I asked on p1654743450371799-slack-wc-rest-api for more context and to verify if this response is expected or not.

- What would be the best approach for this potential workaround? I'd like more input here, but some ideas: 
1 - We could use data we already have when initializing the Refund in order to know if we have an automated payment or not. 
 For example, could we fill this property with the Order payment data? Should we even try build it like this if the property happens to not be in the API response by design?
2 - We could to use some sort of data force-refresh (we go back to the idea of updating the Order asynchronously at some point)

**Comment by joshheald** [comment]
> Is it expected that the API POST request does not contain refunded_payment at this point? This could be a bug or a feature, so we would want either a temporary or a permanent workaround. For the moment I asked on p1654743450371799-slack-wc-rest-api for more context and to verify if this response is expected or not.

I think it's a bug. Supporting evidence: [the docs do include refunded_payment in the POST response example](https://woocommerce.github.io/woocommerce-rest-api-docs/#create-a-refund)

> What would be the best approach for this potential workaround? I'd like more input here, but some ideas: 
> 1 - We could use data we already have when initializing the Refund in order to know if we have an automated payment or not.
> For example, could we fill this property with the Order payment data? Should we even try build it like this if the property happens to not be in the API response by design?
> 2 - We could to use some sort of data force-refresh (we go back to the idea of updating the Order asynchronously at some point)

Both of these might work, but number 1 does feel like quite a bit of a hack to me. 

I think it would be safest to do a [GET request for the `/wp-json/wc/v3/orders/<id>/refunds/<refund_id>`](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-a-refund) and put the result in our core data store, directly after the successful POST. That _should_ mean that we are using the latest data. This does mean additional load on the API, but performing refunds is not a high-volume action, so I don't think we should be too concerned about that.

We could mark this with a note to remove it when the API issue is fixed and some threshold of updates is reached, and add an iOS Github issue to the same effect as well, to mention in the TODO.
